syntax = "proto3";

package juncore.protocol;

// 기본 메시지 타입
enum MessageType {
  UNKNOWN = 0;
  LOGIN_REQUEST = 1;
  LOGIN_RESPONSE = 2;
  CHAT_MESSAGE = 3;
  PLAYER_MOVE = 4;
  PLAYER_POSITION = 5;
  PING = 6;
  PONG = 7;
}

// 로그인 요청
message LoginRequest {
  string username = 1;
  string password = 2;
  string version = 3;
}

// 로그인 응답
message LoginResponse {
  bool success = 1;
  string message = 2;
  int32 player_id = 3;
}

// 채팅 메시지
message ChatMessage {
  int32 player_id = 1;
  string message = 2;
  int64 timestamp = 3;
}

// 플레이어 이동
message PlayerMove {
  int32 player_id = 1;
  float x = 2;
  float y = 3;
  float z = 4;
  float rotation = 5;
}

// 플레이어 위치
message PlayerPosition {
  int32 player_id = 1;
  float x = 2;
  float y = 3;
  float z = 4;
  float rotation = 5;
  int64 timestamp = 6;
}

// 패킷 헤더
message PacketHeader {
  MessageType type = 1;
  int32 size = 2;
  int64 timestamp = 3;
}

// 통합 게임 패킷
message GamePacket {
  PacketHeader header = 1;
  oneof payload {
    LoginRequest login_request = 2;
    LoginResponse login_response = 3;
    ChatMessage chat_message = 4;
    PlayerMove player_move = 5;
    PlayerPosition player_position = 6;
  }
}